---
title:    "Epigenetic dynamics during capacitation of naïve human pluripotent stem cells"
subtitle: "Promoter and gene expression analysis"
author:   "João Pedro Agostinho de Sousa"
date:     "May 18, 2021"
output:
  html_document:
    df_print: paged
    toc: TRUE
    #toc_float: TRUE
    theme: default
editor_options:
  chunk_output_type: console
---

# Dependencies
```{r Packages, results = "hide", warning = FALSE}

purrr::walk(list.files("analysis/functions", pattern = "load_project_packages", full.names = T), source)

```

# Loading promoter counts
```{r}

load("output/database/counts_in_promoters_full.RData")

counts_in_promoters_full <- unique(counts_in_promoters_full)

```

# Loading RNA-seq clusters
```{r}

load("output/database/rnaseq_DGE_clusters.RData")

```

# RNA-seq clusters by gene
```{r}

# Heatmap colors
heatmap_colors_scaled <- circlize::colorRamp2(seq(-2, 2, 4/10), c(rev(brewer.pal(n = 11, name = "RdBu"))))

# Heatmap with scaled values
heatmap_scaled <- 
  mfuzz_clusters %>%
  select(-gene_id) %>%
  select(-CLUSTER) %>%
  as.matrix() %>%
  .[, c("0", "1", '2', '3', '7', "10", "20X", "20E", "H9 Primed")] %>%
  set_rownames(c(1:nrow(.))) %>% 
  t %>% 
  scale %>%
  t %>% 
  ComplexHeatmap::Heatmap(name             = "Row Z-score",
                          col              = heatmap_colors_scaled,
                          row_split        = mfuzz_clusters$CLUSTER,
                          border           = T,
                          cluster_columns  = F,
                          cluster_rows     = F, 
                          show_row_names   = F,
                          column_names_rot = 60, 
                          row_names_side   = "left")

```

# RNA-seq clusters by average expression
```{r}

# Colors for RNA-Seq heatmap
heatmap_colors <- circlize::colorRamp2(c(-1.5, -1, -0.5, 0, 0.5, 1, 1.5), c(rev(brewer.pal(n = 11, name = "RdBu")[c(2,3,5,6,7,10,11)])))

# RNA-Seq heatmap
rnaseq_ht <- 
  mfuzz_clusters %>%
  filter(gene_id %in% unique(counts_in_promoters_full$gene_id)) %>%
  gather(group, rnaseq, -gene_id, -CLUSTER) %>%
  group_by(group, CLUSTER) %>% 
  summarise(rnaseq = mean(rnaseq, na.rm = T)) %>% 
  ungroup %>% 
  spread(group, rnaseq) %>%  
  arrange(CLUSTER) %>% 
  column_to_rownames("CLUSTER") %>%
  as.matrix() %>% 
  .[, c("0", "10", "20X", "20E", "H9 Primed")] %>% 
  ComplexHeatmap::Heatmap(matrix               = .,
                          col                  = rev(brewer.pal(n = 11, name = "RdBu")),
                          name                 = "RNA-Seq",
                          cluster_rows         = F, 
                          show_row_names       = F, 
                          column_names_rot     = 60,
                          cluster_columns      = F,
                          heatmap_legend_param = list(legend_direction = "horizontal"),
                          column_title         = "RNA-Seq",
                          rect_gp              = gpar(color = "black"),
                          border               = T)

# Colors for epigenetic modifications heatmap
epi_modifications_colors <- circlize::colorRamp2(seq(-0.40, 0.40, 0.125), c(rev(brewer.pal(n = 11, name = "RdBu")[c(2,3,5,6,7,10,11)])))
methylation_colors <- circlize::colorRamp2(seq(0, 1, 0.1), c(rev(brewer.pal(n = 11, name = "RdBu")[c(1,1,2:10)])))

# Epigenetic modifications heatmap
epi_modifications_ht <-
map(c("ATAC-Seq", "H3K27ac", "H3K27me3", "H3K4me1", "H3K4me3", "H3K9me3", "Methylation") %>% set_names(.,.),
  ~ counts_in_promoters_full %>%
    left_join(unique(RNAseq_metadata[, c("day", "group")])) %>% 
    mutate(group = str_remove(group, "-EOS") %>% str_replace("H9$", "H9 Primed")) %>% 
    mutate(day = str_remove(day, "-EOS") %>% str_replace("H9$", "H9 Primed")) %>%
    inner_join(unique(mfuzz_clusters[, c("gene_id", "CLUSTER")])) %>% 
    select(c("gene_id", "day", .x, "CLUSTER")) %>%
    group_by(day, CLUSTER) %>% 
    summarise(!!.x := mean(!!sym(.x), na.rm = T)) %>% 
    ungroup %>%
    spread(day, !!sym(.x)) %>%  
    arrange(CLUSTER) %>% 
    column_to_rownames("CLUSTER") %>%
    as.matrix() %>% 
    .[, c("0", "10", "20X", "20E", "H9 Primed")] %>%
    ComplexHeatmap::Heatmap(matrix               = .,
                            col                  = if(.x == "Methylation"){methylation_colors}else{rev(brewer.pal(n = 11, name = "RdBu"))},
                            name                 = .x,
                            cluster_rows         = F, 
                            show_row_names       = F,
                            cluster_columns      = F, 
                            heatmap_legend_param = list(legend_direction = "horizontal"),
                            column_names_rot     = 60,
                            column_title         = .x,
                            rect_gp              = gpar(color = "black"),
                            border               = T)
    )

# Combinations of RNAs-seq and epigenetic modifications heatmap
ht_rnaseq_epi <- 
  rnaseq_ht +
  epi_modifications_ht$`ATAC-Seq` +
  epi_modifications_ht$H3K4me3 + 
  epi_modifications_ht$H3K27ac +
  epi_modifications_ht$H3K4me1 + 
  epi_modifications_ht$H3K27me3 + 
  epi_modifications_ht$H3K9me3 + 
  epi_modifications_ht$Methylation

# Number of genes in each clusters
counts_in_promoters_full %>%
    left_join(unique(RNAseq_metadata[, c("day", "group")])) %>% 
    mutate(group = str_remove(group, "-EOS") %>% str_replace("H9$", "H9 Primed")) %>% 
    mutate(day = str_remove(day, "-EOS") %>% str_replace("H9$", "H9 Primed")) %>%
    inner_join(unique(mfuzz_clusters[, c("gene_id", "CLUSTER")])) %>% 
    select(gene_id,CLUSTER, CGI) %>% 
    unique %>% named_group_split(CLUSTER) %>% map(~ pull(.x, gene_id) %>% unique %>% length)

```

# Gene ontology between clusters
```{r}

GO_clusters <-
  compareCluster(mfuzz_clusters %>% named_group_split(CLUSTER) %>% map(~ pull(.x, gene_id)), 
                 fun           = "enrichGO", 
                 OrgDb         = "org.Hs.eg.db",
                 keyType       = 'ENSEMBL',
                 ont           = "ALL",
                 pAdjustMethod = "BH",
                 qvalueCutoff  = 0.1,
                 pvalueCutoff  = 0.1)

GO_clusters_plot <-
  clusterProfiler::dotplot(GO_clusters, group = FALSE, showCategory = 4) +
  scale_color_distiller(palette = "Reds", direction = 1) +
  scale_y_discrete(labels = function(x) str_wrap(str_to_sentence(x), width = 100)) +
  labs(color = "Adj. p-value", size = "Gene ratio") +
  theme(panel.grid.major = element_line(linetype = 2)) +
  scale_color_gradient(low = "#f75656", high = "#6b1313")

```

# Saving plots
```{r}

svg(filename = "output/figures/revision/mfuzz_clusters_v2_heatmap_scaled.svg", width = 3, height = 6)
pdf(file = "output/figures/revision/mfuzz_clusters_v2_heatmap_scaled.pdf", width = 3, height = 6)
heatmap_scaled
dev.off()

svg(filename = "output/figures/revision/promoter_expression_cluster_1.svg", width = 17, height = 3.5)
pdf(file = "output/figures/revision/promoter_expression_cluster_1.pdf", width = 17, height = 3.5)
ht_rnaseq_epi
dev.off()

svg(filename = "output/figures/revision/clusters_GO.svg", width = 9, height = 7.5)
pdf(file = "output/figures/revision/clusters_GO.pdf", width = 9, height = 7.5)
GO_clusters_plot
dev.off()

```

# Session information
```{r session_info, echo = FALSE}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.