---
title:    "Epigenetic dynamics during capacitation of naïve human pluripotent stem cells"
subtitle: "Correlation between epigenetic marks"
author:   "João Pedro Agostinho de Sousa"
date:     "Date: `r format(Sys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    #toc_float: TRUE
    theme: default
editor_options:
  chunk_output_type: console
---

# Dependencies
```{r Packages, results = "hide", warning = FALSE}

purrr::walk(list.files("analysis/functions", pattern = "load_project_packages", full.names = T), source)

```

# Loading data
```{r}

# Loading ChIP-seq, ATAC-seq, and PBAT read counts
load(file = "output/database/counts_in_windows_2Kb.RData")

```

# Loading counts inside unbiased regions
```{r}

# Methylation ratio
PBAT_meth_ratio <- 
  PBAT_meth_ratio %>%
  .[, !str_detect(colnames(.), "d3")] %>% 
  .[!str_detect(rownames(.), "X:|Y:"), ] %>%
  as.data.frame() %>% 
  rownames_to_column("position") %>%
  gather(sample_id, meth_perc, -position) %>%
  left_join(unique(PBAT_metadata[, c("sample_id", "group")]), by = "sample_id") %>%
  group_by(group, position) %>%
  summarise(Methylation = mean(meth_perc, na.rm = TRUE)) %>%
  ungroup


# ChIP-seq and ATAC-seq counts
plan("multisession", workers = 10)

chipseq_atacseq_sample_counts <- 
  chipseq_atacseq_transformed_counts %>% 
  future_map2(names(.),
              ~ assay(.x) %>%
                .[, !str_detect(colnames(.), "d3")] %>% 
                .[!str_detect(rownames(.), "X:|Y:"), ] %>%
                as.data.frame() %>%
                rownames_to_column("position") %>%
                gather(sample_id, value, -position) %>%
                left_join(unique(RNAseq_metadata[, c("sample_id", "group")]), by = "sample_id") %>%
                group_by(group, position) %>%
                summarise(!!.y := mean(value, na.rm = TRUE)) %>%
                ungroup)

plan("sequential")

```

# Correlation between epigenetic marks
```{r}

seqtype_correlation <- 
  chipseq_atacseq_sample_counts %>% 
  purrr::reduce(full_join) %>% 
  full_join(PBAT_meth_ratio, by = c("position", "group")) %>%
  named_group_split(group) %>%
  map2_dfr(names(.),
           ~ select(.x, -group, -position) %>%
             na.omit() %>% 
             cor(method = "spearman") %>% 
             as.data.frame %>% 
             rownames_to_column("group_1") %>% 
             gather(group_2, correlation, -group_1) %>% 
             mutate(group = .y)
           ) %>% 
  mutate(group_1 = factor(group_1, levels = c("ATAC-Seq",
                                              "H3K4me3",
                                              "H3K27ac",
                                              "H3K4me1",
                                              "H3K27me3",
                                              "H3K9me3",
                                              "Methylation")),
         group_2 = factor(group_2, levels = c("ATAC-Seq",
                                              "H3K4me3",
                                              "H3K27ac",
                                              "H3K4me1",
                                              "H3K27me3",
                                              "H3K9me3",
                                              "Methylation")),
         group = factor(group, levels = c(
                                              "HNES1 d0",
                                              "cR-H9-EOS d0",
                                              "HNES1 d3",
                                              "cR-H9-EOS d3",
                                              "HNES1 d10",
                                              "cR-H9-EOS d10",
                                              "cR-H9-EOS d20X", 
                                              "cR-H9-EOS d20E",
                                              "H9-EOS"
                                              )))

# Changing group names
seqtype_correlation$group <- seqtype_correlation$group %>% str_replace("^H9-EOS$", "H9 Primed") %>% str_remove("-EOS")
seqtype_correlation       <- seqtype_correlation %>% mutate(group = factor(group, levels = c("HNES1 d0", "cR-H9 d0", "HNES1 d10", "cR-H9 d10", "cR-H9 d20X", "cR-H9 d20E", "H9 Primed")))

```

# Correlation plot
```{r}

seqtype_correlation_plot <- 
  seqtype_correlation %>%
  ggplot(aes(group, fct_rev(group_1), label = round(correlation, 2), fill = correlation)) +
  geom_tile(color = "black") +
  geom_text(size = 2.5) +
  scale_fill_stepsn(breaks = seq(-0.2, 1, 0.2), colours = rev(brewer.pal(n = 11, name = "RdBu")[c(1,2,3,4,5,5,7,10,11,11,11)]), limits = c(-0.20, 1), values = scales::rescale(seq(-0.9, 1, 0.1), from = c(-0.5, 1))) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0,0)) +
  labs(fill = "Spearman correlation", y = "", x = "") +
  theme(axis.ticks       = element_blank(),
        axis.text.x      = element_text(angle = 45, hjust = 1, vjust = 1, size = 10, color = "black"),
        axis.text.y      = element_text(size = 10, color = "black"),
        strip.text       = element_text(face = "bold", size = 15),
        strip.background = element_blank(),
        legend.position  = "bottom") +
  guides(fill = guide_colourbar(frame.colour   = "Black", 
                                ticks.colour   = "black", 
                                direction      = "horizontal", 
                                title.position = "left", 
                                barwidth       = 10, 
                                barheight      = 1)) +
  facet_grid(~ group_2)

```

# Saving plots
```{r}

svg(filename = "output/figures/02_genome_wide_analysis/seqtype_correlation_plot.svg", width = 16, height = 4)
#pdf(file = "output/figures/02_genome_wide_analysis/seqtype_correlation_plot.pdf", width = 16, height = 4)
seqtype_correlation_plot
dev.off()

```

# Session information
```{r session_info, echo = FALSE}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.