---
title:    "Epigenetic dynamics during capacitation of naïve human pluripotent stem cells"
subtitle: "Similarity between sample groups"
author:   "João Pedro Agostinho de Sousa"
date:     "Date: `r format(Sys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    #toc_float: TRUE
    theme: default
editor_options:
  chunk_output_type: console
---

# Dependencies
```{r Packages, results = "hide", warning = FALSE}

purrr::walk(list.files("analysis/functions", pattern = "load_project_packages", full.names = T), source)

```

# Loading data
```{r}

# Loading ChIP-seq, ATAC-seq, and PBAT read counts
load(file = "output/database/counts_in_windows_2Kb.RData")

```

# Correlation between groups
```{r}

PBAT_meth_ratio <- 
  PBAT_meth_ratio %>%
  .[, !str_detect(colnames(.), "d3")] %>% 
  .[!str_detect(rownames(.), "X:|Y:"), ] %>%
  as.data.frame() %>% 
  na.omit() %>% 
  rownames_to_column("position") %>%
  gather(sample_id, meth_perc, -position) %>%
  left_join(unique(PBAT_metadata[, c("sample_id", "group")]), by = "sample_id") %>%
  group_by(group, position) %>%
  summarise(Methylation = mean(meth_perc)) %>%
  ungroup

# Methylation correlation
PBAT_meth_correlation <- 
  spread(PBAT_meth_ratio, group, Methylation) %>% 
  select(-position) %>% 
  as.matrix() %>%
  cor(method = "pearson") %>% 
  as.data.frame() %>% 
  rownames_to_column("group_1") %>% 
  gather(group_2, correlation, -group_1) %>% 
  mutate(group_1 = str_replace(group_1, "_", " "),
         group_2 = str_replace(group_2, "_", " ")) %>% 
  mutate(group_1 = factor(group_1, levels = c("HNES1 d0", "cR-H9-EOS d0", "HNES1 d3", "cR-H9-EOS d3", "HNES1 d10", "cR-H9-EOS d10", "cR-H9-EOS d20X",  "cR-H9-EOS d20E", "H9-EOS")),
         group_2 = factor(group_2, levels = c("HNES1 d0", "cR-H9-EOS d0", "HNES1 d3", "cR-H9-EOS d3", "HNES1 d10", "cR-H9-EOS d10", "cR-H9-EOS d20X",  "cR-H9-EOS d20E", "H9-EOS"))
         )

# ChIP-Seq and ATAC-Seq correction
chipseq_atacseq_group_correlation <- 
  map(chipseq_atacseq_transformed_counts,
  ~ assay(.x) %>%
    .[, !str_detect(colnames(.), "d3")] %>% 
    .[!str_detect(rownames(.), "X:|Y:"), ] %>%
    as.data.frame() %>% 
    rownames_to_column("position") %>%
    gather(sample_id, value, -position) %>%
    left_join(unique(RNAseq_metadata[, c("sample_id", "group")]), by = "sample_id") %>%
    group_by(group, position) %>%
    summarise(mean_value = mean(value)) %>%
    ungroup %>%
    spread(key = group, value = mean_value) %>%
    select(-position) %>%
    cor(method = "pearson") %>%
    as.data.frame() %>% 
    rownames_to_column("group_1") %>% 
    gather(group_2, correlation, -group_1) %>% 
    mutate(group_1 = factor(group_1, levels = c("HNES1 d0", "cR-H9-EOS d0", "HNES1 d3", "cR-H9-EOS d3", "HNES1 d10", "cR-H9-EOS d10", "cR-H9-EOS d20X",  "cR-H9-EOS d20E", "H9-EOS")),
           group_2 = factor(group_2, levels = c("HNES1 d0", "cR-H9-EOS d0", "HNES1 d3", "cR-H9-EOS d3", "HNES1 d10", "cR-H9-EOS d10", "cR-H9-EOS d20X",  "cR-H9-EOS d20E", "H9-EOS"))
    ))

group_correlation <- 
list(
chipseq_atacseq_group_correlation %>% map2(names(.), ~ mutate(.x, seqtype = .y)) %>% purrr::reduce(rbind),
PBAT_meth_correlation %>% mutate(seqtype = "Methylation")
) %>% 
  purrr::reduce(rbind) %>%
  mutate(group_1_cell = case_when(
    
    str_detect(group_1, "cR-H9-EOS") ~ "cR-H9-EOS",
    str_detect(group_1, "^H9-EOS") ~ "H9-EOS",
    str_detect(group_1, "HNES1") ~ "HNES1"
    
  ),
  group_2_cell = case_when(
    
    str_detect(group_2, "cR-H9-EOS") ~ "cR-H9-EOS",
    str_detect(group_2, "^H9-EOS") ~ "H9-EOS",
    str_detect(group_2, "HNES1") ~ "HNES1"
    
  ),
  seqtype = factor(seqtype, levels = c("ATAC-Seq", "H3K4me3", "H3K27ac", "H3K4me1", "H3K27me3", "H3K9me3", "Methylation")))

group_correlation_plot <- 
group_correlation %>%
ggplot(aes(group_1, group_2, label = round(correlation, 2), fill = correlation)) +
  geom_tile(color = "black") +
  geom_text(size = 3) +
  scale_fill_gradientn(colours = rev(brewer.pal(11, "RdBu"))) +
  scale_y_discrete(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  labs(fill = "Correlation", y = "", x = "") +
  theme(axis.ticks       = element_blank(),
        axis.text.x      = element_text(angle = 45, hjust = 1, vjust = 1, face = "bold", size = 10),
        axis.text.y      = element_text(face = "bold", size = 10),
        strip.text       = element_text(face = "bold", size = 15),
        strip.background = element_blank()) +
  guides(fill = guide_colourbar(frame.colour = "Black", ticks.colour = "black")) +
  facet_grid( ~ seqtype)

```

# Similarity plots
```{r}

similarity_plot_data <- 
list(
group_correlation %>% filter(group_2 == "cR-H9-EOS d0" & str_detect(group_1, "H9-EOS")),
group_correlation %>% filter(group_2 == "cR-H9-EOS d10" & str_detect(group_1, "H9-EOS")),
group_correlation %>% filter(group_2 == "cR-H9-EOS d20X" & str_detect(group_1, "H9-EOS")),
group_correlation %>% filter(group_2 == "cR-H9-EOS d20E" & str_detect(group_1, "H9-EOS")),
group_correlation %>% filter(group_2 == "H9-EOS" & str_detect(group_1, "H9-EOS")),
group_correlation %>% filter(group_2 == "HNES1 d0" & str_detect(group_1, "HNES1|^H9-EOS")),
group_correlation %>% filter(group_2 == "HNES1 d10" & str_detect(group_1, "HNES1|^H9-EOS"))
) %>%
  purrr::reduce(rbind) %>%
  mutate(group_1 = str_replace(group_1, "cR-H9-EOS d|HNES1 d", "") %>% factor(levels = c("0", "3", "10", "20X", "20E", "H9-EOS"))) %>% 
  mutate(seqtype = factor(seqtype, levels = c("ATAC-Seq", "H3K4me3", "H3K27ac", "H3K4me1", "H3K27me3", "H3K9me3", "Methylation")))

# Changing group names
similarity_plot_data$group_1      <- similarity_plot_data$group_1 %>% str_replace("^H9-EOS$", "H9 Primed") %>% str_remove("-EOS")
similarity_plot_data              <- similarity_plot_data %>% mutate(group_1 = factor(group_1, levels = c("0", "10", "20X", "20E", "H9 Primed")))
    
similarity_plot_data$group_2      <- similarity_plot_data$group_2 %>% str_replace("^H9-EOS$", "H9 Primed") %>% str_remove("-EOS")
similarity_plot_data              <- similarity_plot_data %>% mutate(group_2 = factor(group_2, levels = c("HNES1 d0", "cR-H9 d0", "HNES1 d10", "cR-H9 d10", "cR-H9 d20X", "cR-H9 d20E", "H9 Primed")))

similarity_plot_data$group_1_cell <- similarity_plot_data$group_1_cell %>% str_replace("^H9-EOS$", "H9 Primed") %>% str_remove("-EOS")
similarity_plot_data              <- similarity_plot_data %>% mutate(group_1_cell = factor(group_1_cell, levels = c("HNES1", "cR-H9", "H9 Primed")))

similarity_plot_data$group_2_cell <- similarity_plot_data$group_2_cell %>% str_replace("^H9-EOS$", "H9 Primed") %>% str_remove("-EOS")
similarity_plot_data              <- similarity_plot_data %>% mutate(group_2_cell = factor(group_2_cell, levels = c("HNES1", "cR-H9", "H9 Primed")))

# Separating H9 Primed similarity comparison with cR-H9 and HNES1 cells
similarity_plot_data <- 
rbind(similarity_plot_data %>% mutate(group_2 = as.character(group_2), group_2 = case_when(group_2 == "H9 Primed" ~ "H9 Primed (cR-H9)", TRUE ~ group_2)),
  rbind(similarity_plot_data %>% filter(group_1_cell == "H9 Primed" & group_2_cell == "HNES1") %>% mutate(group_1 = str_extract(group_2, "d\\d+") %>% str_remove("d")) %>% mutate(group_2 = "H9 Primed (HNES1)"),
      similarity_plot_data %>% filter(group_1_cell == "H9 Primed" & group_2_cell == "H9 Primed") %>% mutate(group_2 = "H9 Primed (HNES1)"))) %>%
  mutate(group_2 = factor(group_2, levels = c("HNES1 d0", "cR-H9 d0", "HNES1 d10", "cR-H9 d10", "cR-H9 d20X", "cR-H9 d20E", "H9 Primed (HNES1)", "H9 Primed (cR-H9)")))

similarity_plot <-
ggplot(rbind(similarity_plot_data, similarity_plot_data %>% group_by(group_1, group_2) %>% mutate(correlation = median(correlation)) %>% ungroup %>% mutate(seqtype = "Median correlation") %>% unique),
       aes(group_1, forcats::fct_rev(seqtype), fill = correlation, label = round(correlation, 2))) +
  geom_tile(color = "black") +
  geom_text(size = 3.5) +
  scale_fill_stepsn(n.breaks = 9, colours = brewer.pal(n = 9, name = "Blues")) +
  scale_y_discrete(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_publication() +
  theme(axis.ticks        = element_blank(),
        axis.text.x       = element_text(size = 12),
        axis.text.y       = element_text(size = 12),
        strip.text        = element_text(face = "bold", size = 15),
        axis.title        = element_text(size = 14),
        strip.background  = element_blank(),
        legend.key.width  = unit(2.5, "cm"),
        legend.key.height = unit(0.4, "cm"),
        legend.position   = "bottom",
        legend.box        = "horizontal") +
  labs(x = "Day", y = "") +
  guides(fill = guide_colourbar(title = "Pearson correlation", title.position = "left", title.hjust = 0.5, barwidth = unit(7, "cm"), barheight = unit(0.30, "cm"), frame.colour = "black", ticks.colour = "black")) +
  facet_grid( ~ group_2, scales = "free_x")

```

# Saving plots
```{r}

svg(filename = "output/figures/02_genome_wide_analysis/similarity_plot_1.svg", width = 18, height = 4)
pdf(file = "output/figures/02_genome_wide_analysis/similarity_plot_1.pdf", width = 18, height = 4)
similarity_plot
dev.off()

```

# Session information
```{r session_info, echo = FALSE}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.