---
title:    "Epigenetic dynamics during capacitation of naïve human pluripotent stem cells"
subtitle: "Correlation in annotated regions between epigenetic marks"
author:   "João Pedro Agostinho de Sousa"
date:     "Date: `r format(Sys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    #toc_float: TRUE
    theme: default
editor_options:
  chunk_output_type: console
---

# Dependencies
```{r Packages, results = "hide", warning = FALSE}

purrr::walk(list.files("analysis/functions", pattern = "load_project_packages", full.names = T), source)

```

# Loading data
```{r}

# Loading ChIP-seq, ATAC-seq, and PBAT read counts
load(file = "output/database/counts_in_windows_2Kb.RData")

```

# Tilling the genome
```{r}

# Getting the chromosome length
chromosome_lenght <-
  read_tsv("genomes/Homo_sapiens/GRCh38/Ensembl/release-98/Homo_sapiens.GRCh38.dna_sm.primary_assembly.chrSize.txt", col_names = F) %>%
  dplyr::rename(Chromosome = X1, Length = X2)

# Creating a vector of the chromosomes lengths
chromosome_lenght_vector <-
  chromosome_lenght$Length %>% setNames(chromosome_lenght$Chromosome)

# Tilling the genome into 2kb windows
tilling_genome_windows_2Kb <-
  tileGenome(chromosome_lenght_vector, tilewidth = 2000) %>% unlist

# Removing windows that overlap centromeres
tilling_genome_windows_2Kb <- subsetByOverlaps(tilling_genome_windows_2Kb, genome_annotation$Centromeres, invert = T)

```

# Annotating probes
```{r}

# Chosing the annotations
genome_annotation_subset <- 
  genome_annotation %>% .[c(
    "CGI Promoters",
    "Non-CGI Promoters",
    "CpG islands",
    "Distal Naïve enhancers",
    "Distal Primed enhancers",
    "Distal Naïve super-enhancers",
    "Distal Primed super-enhancers",
    "Exons",
    "Introns",
    "Repeats",
    "Intergenic regions",
    "CGI shores",
    "CGI shelves",
    "CGI Promoters (high expression)",
    "CGI Promoters (median expression)",
    "CGI Promoters (low expression)",
    "Non-CGI Promoters (high expression)",
    "Non-CGI Promoters (median expression)",
    "Non-CGI Promoters (low expression)"
    )] %>% GRangesList()

# Annotated seqmonk probes as a GRanges objects
windows_2Kb_autosomes_annotated_gr <- 
 tilling_genome_windows_2Kb %>% 
  .[seqnames(.) %in% 1:22] %>% 
 annotate_regions(genomic_regions             = ., 
                  annotation_list             = genome_annotation_subset,
                  with_duplicated_annotations = TRUE,
                  apply_GRanges_reduce        = FALSE) %>% 
  named_group_split(annotation) %>% 
  map(~ makeGRangesFromDataFrame(.x, keep.extra.columns = T))

# Annotated seqmonk probes as a dataframe
windows_2Kb_autosomes_annotated_df <- 
 tilling_genome_windows_2Kb %>%
 .[seqnames(.) %in% 1:22] %>% 
 annotate_regions(genomic_regions             = ., 
                  annotation_list             = genome_annotation_subset,
                  with_duplicated_annotations = TRUE,
                  apply_GRanges_reduce        = FALSE) %>% 
  mutate(position = paste(seqnames, start, end, sep = ":")) %>%
  select(position, annotation) %>%
  na.omit %>% 
  unique

```

# Loading counts inside unbiased regions
```{r}

# Methylation ratio
PBAT_meth_ratio <- 
  PBAT_meth_ratio %>%
  .[, !str_detect(colnames(.), "d3")] %>% 
  .[!str_detect(rownames(.), "X:|Y:"), ] %>%
  as.data.frame() %>% 
  rownames_to_column("position") %>%
  gather(sample_id, meth_perc, -position) %>%
  left_join(unique(PBAT_metadata[, c("sample_id", "group")]), by = "sample_id") %>%
  group_by(group, position) %>%
  summarise(Methylation = mean(meth_perc, na.rm = TRUE)) %>%
  ungroup

# ChIP-seq and ATAC-seq counts
plan("multisession", workers = 20)

chipseq_atacseq_sample_counts <- 
  chipseq_atacseq_transformed_counts %>% 
  future_map2(names(.),
              ~ assay(.x) %>%
                .[, !str_detect(colnames(.), "d3")] %>% 
                .[!str_detect(rownames(.), "X:|Y:"), ] %>%
                as.data.frame() %>%
                rownames_to_column("position") %>%
                gather(sample_id, value, -position) %>%
                left_join(unique(RNAseq_metadata[, c("sample_id", "group")]), by = "sample_id") %>%
                group_by(group, position) %>%
                summarise(!!.y := mean(value, na.rm = TRUE)) %>%
                ungroup)

plan("sequential")

```

# PBAT, ChIP-Seq, and ATAC-Seq counts by annotation
```{r}

PBAT_meth_ratio_by_annotation <- map(unique(windows_2Kb_autosomes_annotated_df$annotation) %>% set_names(.,.), ~ PBAT_meth_ratio %>% .[.$position %in% windows_2Kb_autosomes_annotated_df[windows_2Kb_autosomes_annotated_df$annotation == .x, "position"],])

chipseq_atacseq_sample_counts_by_annotation <- map(unique(windows_2Kb_autosomes_annotated_df$annotation) %>% set_names(.,.), function(x){map(chipseq_atacseq_sample_counts, ~ .x %>% .[.$position %in% windows_2Kb_autosomes_annotated_df[windows_2Kb_autosomes_annotated_df$annotation == x, "position"],])})

```

# Correlation between epigenetic marks
```{r}

seqtype_correlation <-
map2(chipseq_atacseq_sample_counts_by_annotation, PBAT_meth_ratio_by_annotation, function(x, y){

  seqtype_correlation <- 
    x %>% 
    purrr::reduce(full_join) %>% 
    full_join(y, by = c("position", "group")) %>%
    named_group_split(group) %>%
    map2_dfr(names(.),
             ~ select(.x, -group, -position) %>%
               na.omit() %>% 
               cor(method = "spearman") %>% 
               as.data.frame %>% 
               rownames_to_column("group_1") %>% 
               gather(group_2, correlation, -group_1) %>% 
               mutate(group = .y)
             ) %>% 
    mutate(group_1 = factor(group_1, levels = c("ATAC-Seq",
                                                "H3K4me3",
                                                "H3K27ac",
                                                "H3K4me1",
                                                "H3K27me3",
                                                "H3K9me3",
                                                "Methylation")),
           group_2 = factor(group_2, levels = c("ATAC-Seq",
                                                "H3K4me3",
                                                "H3K27ac",
                                                "H3K4me1",
                                                "H3K27me3",
                                                "H3K9me3",
                                                "Methylation")),
           group = factor(group, levels = c(
                                                "HNES1 d0",
                                                "cR-H9-EOS d0",
                                                "HNES1 d3",
                                                "cR-H9-EOS d3",
                                                "HNES1 d10",
                                                "cR-H9-EOS d10",
                                                "cR-H9-EOS d20X", 
                                                "cR-H9-EOS d20E",
                                                "H9-EOS"
                                                )))
  
  # Changing group names
  seqtype_correlation$group <- seqtype_correlation$group %>% str_replace("^H9-EOS$", "H9 Primed") %>% str_remove("-EOS")
  seqtype_correlation       <- seqtype_correlation %>% mutate(group = factor(group, levels = c("HNES1 d0", "cR-H9 d0", "HNES1 d10", "cR-H9 d10", "cR-H9 d20X", "cR-H9 d20E", "H9 Primed")))
  
  return(seqtype_correlation)
})

```

# Merging correlation results between annotated regions
```{r}

seqtype_correlation_w_annotation <- seqtype_correlation %>% map2_dfr(names(.), ~ mutate(.x, annotation = .y))

```

# Correlation plot
```{r}

seqtype_correlation_by_annotation_plot <- 
  seqtype_correlation_w_annotation %>%
  mutate(annotation = factor(annotation, levels = 
  c(
    "CpG islands",
    "CGI shores",
    "CGI shelves",
    "CGI Promoters",
    "CGI Promoters (high expression)",
    "CGI Promoters (median expression)",
    "CGI Promoters (low expression)",
    "Distal Naïve enhancers",
    "Distal Naïve super-enhancers",
    "Distal Primed enhancers",
    "Distal Primed super-enhancers",
    "Non-CGI Promoters",
    "Non-CGI Promoters (high expression)",
    "Non-CGI Promoters (median expression)",
    "Non-CGI Promoters (low expression)",
    "Exons",
    "Introns",
    "Repeats",
    "Intergenic regions"
    ))) %>%
  ggplot(aes(group, fct_rev(annotation), label = round(correlation, 2), fill = correlation)) +
  geom_tile(color = "black", size = 0.1) +
  scale_fill_stepsn(breaks = seq(-1,1,0.2), colours = rev(brewer.pal(n = 11, name = "RdBu")[c(1,1,3,4,5,6,7,8,9,11,11)]), limits = c(-1, 1)) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0,0)) +
  labs(fill = "Spearman correlation", y = "", x = "") +
  theme(axis.ticks       = element_blank(),
        axis.text.x      = element_text(angle = 45, hjust = 1, vjust = 1, size = 8, color = "black"),
        axis.text.y      = element_text(size = 7, color = "black"),
        strip.text       = element_text(face = "bold", size = 15),
        strip.background = element_blank(), 
        panel.border     = element_rect(colour = "black", fill = NA, size = 0.25),
        legend.position  = "bottom") +
  guides(fill = guide_colourbar(frame.colour   = "Black", 
                                ticks.colour   = "black", 
                                direction      = "horizontal", 
                                title.position = "left", 
                                barwidth       = 10, 
                                barheight      = 1)) +
  facet_grid(group_1 ~ group_2)

```

# Saving plots
```{r}

svg(filename = "output/figures/02_genome_wide_analysis/seqtype_correlation_by_annotation_plot.svg", width = 14, height = 13.5)
#pdf(file = "output/figures/02_genome_wide_analysis/seqtype_correlation_by_annotation_plot.pdf", width = 14, height = 13.5)
seqtype_correlation_by_annotation_plot
dev.off()

```

# Session information
```{r session_info, echo = FALSE}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.