---
title:    "Epigenetic dynamics during capacitation of naïve human pluripotent stem cells"
subtitle: "Bivalency - heatmap"
author:   "João Pedro Agostinho de Sousa"
date:     "Date: `r format(sSys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    #toc_float: TRUE
    theme: default
editor_options:
  chunk_output_type: console
---

# Dependencies
```{r Packages, results = "hide", warning = FALSE}

purrr::walk(list.files("analysis/functions", pattern = "load_project_packages", full.names = T), source)

```

# Load the saved data
```{r}

# RNA-Seq counts
load("output/database/rnaseq_counts.RData")

# Promoter ChIP-seq and ATAC-seq counts
load("output/database/chipseq_atacseq_counts_in_promoters.RData")

# Promoter Methylation
load("output/database/meth_ratio_in_annotated_regions.RData")
PBAT_meth_ratio <- meth_ratio_in_annotated_regions$Promoters

# ChIP-Seq and ATAC-Seq peaks
load("output/database/chipseq_atacseq_peaks.RData")

```

# Selecting data only in autosomes
```{r}

macs_peaks_gr_in_autosomes <- macs_peaks_gr %>% .[seqnames(.) %in% 1:22]

TSS_in_autosomes <- genome_annotation$Promoters %>% .[seqnames(.) %in% 1:22]

```

# Categorizing promoters based on Methylation percentage
```{r}

PBAT_meth_ratio_by_group <- 
  PBAT_meth_ratio %>%
  left_join(genome_annotation$Promoters %>% as.data.frame() %>% select(seqnames, start, end, width, strand, gene_id)) %>% 
  select(-seqnames, -start, -end, -strand, -width) %>% 
  gather(sample_id, meth_ratio, -gene_id) %>%
  left_join(unique(PBAT_metadata[,c("sample_id", "group")])) %>%
  group_by(gene_id, group) %>% 
  summarize(meth_ratio = mean(meth_ratio, na.rm = T)) %>% 
  ungroup %>% 
  mutate(Methylation = case_when(meth_ratio > 0.5 ~ 1, TRUE ~ 0)) %>% 
  select(gene_id, group, Methylation)

```

# Bivalent promoters
```{r}

# index of the promoters that overlap with peaks 
peaks_TSS_overlaps <- 
map(macs_peaks_gr_in_autosomes$group %>% unique,
    function(x){
      map(macs_peaks_gr_in_autosomes$seqtype %>% unique,
          ~ findOverlaps(
            TSS_in_autosomes,
            macs_peaks_gr_in_autosomes[macs_peaks_gr_in_autosomes$seqtype == .x & macs_peaks_gr_in_autosomes$group == x]) %>%
            as.data.frame() %>%
            pull(queryHits) %>%
            unique
          ) %>%
        set_names(macs_peaks_gr_in_autosomes$seqtype %>% unique)
      }
    ) %>% set_names(macs_peaks_gr_in_autosomes$group %>% unique)

# selecting promoters that overlap with peaks 
peaks_in_TSS <- 
map2(peaks_TSS_overlaps,
     names(peaks_TSS_overlaps),
     function(x, y){
       x %>%
       map2(names(.),
            ~ TSS_in_autosomes %>%
              as_tibble() %>% 
              .[, "gene_id", drop = F] %>% 
              rownames_to_column("index") %>% 
              mutate(!!.y := case_when(index %in% .x ~ 1, TRUE ~ 0))
            ) %>%
         purrr::reduce(full_join) %>%
         gather(seqtype, overlap, -index, -gene_id)
       } %>%
       dplyr::rename(!!y := overlap)
     ) %>%
  purrr::reduce(full_join) %>%
  select(-index) %>%
  gather(group, overlap, -seqtype, -gene_id)

# categorizing promoters by overlapping with peaks
peaks_in_TSS_by_type <- 
  peaks_in_TSS %>%
  named_group_split(group) %>%
  map(~ .x %>% spread(seqtype, overlap) %>% 
        left_join(PBAT_meth_ratio_by_group, by = c("gene_id", "group")) %>% 
        mutate(promoter_type  = case_when(`ATAC-Seq`    == 1 ~ "ATAC-Seq"),
               promoter_type  = case_when(`H3K4me3`     == 1  ~ paste(promoter_type, "H3K4me3", sep = ", "), TRUE ~ promoter_type),
               promoter_type  = case_when(`H3K27ac`     == 1  ~ paste(promoter_type, "H3K27ac", sep = ", "), TRUE ~ promoter_type),
               promoter_type  = case_when(`H3K4me1`     == 1  ~ paste(promoter_type, "H3K4me1", sep = ", "), TRUE ~ promoter_type),
               promoter_type  = case_when(`H3K27me3`    == 1  ~ paste(promoter_type, "H3K27me3", sep = ", "),TRUE ~ promoter_type),
               promoter_type  = case_when(`H3K9me3`     == 1  ~ paste(promoter_type, "H3K9me3", sep = ", "), TRUE ~ promoter_type),
               promoter_type  = case_when(`Methylation` == 1  ~ paste(promoter_type, "Methylation", sep = ", "), TRUE ~ promoter_type),
               promoter_type  = str_remove(promoter_type, "NA, ")) 
      )

# Adding a category column based on the promoter overlap with H3K4me3 and H3K27me3
peaks_in_TSS_by_category <- 
  peaks_in_TSS_by_type %>%
  map(
  ~ .x %>% 
  mutate(promoter_category = case_when(
  
  str_detect(promoter_type, "H3K4me3") & str_detect(promoter_type, "H3K27me3", negate = T) ~ "H3K4me3",
  str_detect(promoter_type, "H3K4me3", negate = T) & str_detect(promoter_type, "H3K27me3") ~ "H3K27me3",
  str_detect(promoter_type, "H3K4me3") & str_detect(promoter_type, "H3K27me3")             ~ "Bivalent",
  TRUE ~ "Other"
  
  ) %>% factor(levels = c("H3K4me3", "Bivalent", "H3K27me3", "Other")))
)

# Changing group names
peaks_in_TSS_by_category <- 
map(peaks_in_TSS_by_category, function(x){
  x$group <- str_replace(x$group, "^H9-EOS$", "H9 Primed") %>% str_remove("-EOS")
  x <- x %>% mutate(group = factor(group, levels = c("HNES1 d0", "cR-H9 d0", "HNES1 d10", "cR-H9 d10", "H9 Primed")))
  return(x)
  })

```

# Heatmap with category clusters, RNA-Seq, H3K4me3, H3K27me3, CGI
```{r}

# Loading promoter counts from RNA-seq, PBAT, ChIP-seq, and ATAC-seq
load("output/database/counts_in_promoters_full.RData")

# Categorizing promoters in each condition
promoters_category_by_group <- 
peaks_in_TSS_by_category %>%
  map2(names(.),
       ~ select(.x, gene_id, promoter_category) %>%
         unique %>%
         dplyr::rename(!!.y := promoter_category)
       ) %>%
  purrr::reduce(left_join)

# Reseted H9 cells heatmap
cRH9EOS_ht <- 
counts_in_promoters_full %>%
filter(group %in% c("cR-H9-EOS d0", "cR-H9-EOS d10")) %>%
select(c("gene_id", "RNA-Seq", "H3K4me3", "H3K27me3", "CGI", "group")) %>%
unique %>% 
left_join(promoters_category_by_group[, c("gene_id", "cR-H9-EOS d0", "cR-H9-EOS d10")] %>%
          arrange(`cR-H9-EOS d0`, `cR-H9-EOS d10`) %>% 
          tidyr::unite("cluster", `cR-H9-EOS d0`:`cR-H9-EOS d10`, sep = "_") %>% 
          mutate(cluster = factor(cluster, levels = unique(cluster))),
          by = "gene_id") %>%
  select(-gene_id) %>%
  mutate(group = factor(group, levels = c("HNES1 d0", "cR-H9-EOS d0", "HNES1 d1", "cR-H9-EOS d1", "HNES1 d2", "cR-H9-EOS d2", "HNES1 d3", "cR-H9-EOS d3", "HNES1 d7", "cR-H9-EOS d7", "HNES1 d10", "cR-H9-EOS d10", "HNES1 d20X", "cR-H9-EOS d20X", "HNES1 d20E", "cR-H9-EOS d20E", "H9-EOS")
  )) %>% 
  filter(group %in% c("cR-H9-EOS d0", "cR-H9-EOS d1", "cR-H9-EOS d2", "cR-H9-EOS d3", "cR-H9-EOS d7", "cR-H9-EOS d10")) %>%
  group_by(group, cluster) %>%
  summarise_all(.funs = ~ mean(.x, na.rm = T)) %>% 
  ungroup() %>%
  gather(seqtype, value, -group, -cluster) %>% 
  named_group_split(seqtype) %>%
  .[c("H3K4me3", "H3K27me3", "CGI", "RNA-Seq")] %>%
  map2(names(.),
       
      function(x, y){
        
      ggplot(x %>% na.omit, aes(group, fct_rev(cluster), fill = value)) +
      geom_tile(color = "black", size = 0.2) +
      scale_fill_gradientn(colours = if(y == "CGI") {rev(brewer.pal(9, "YlGnBu"))} else {rev(brewer.pal(11, "RdBu"))}) +
      theme_publication() +
      theme(axis.title.x    = element_blank(),
            axis.title.y    = element_blank(),
            axis.text.y     = if(y == "H3K4me3") {element_text(hjust = 1)} else {element_blank()},
            axis.ticks.y    = element_blank(),
            axis.text.x     = element_text(angle = 45, hjust = 1, vjust = 1),
            panel.border    = element_blank(),
            #strip.text.y    = if(y == "Methylation") {element_text()} else {element_blank()},
            legend.position = "bottom",
            legend.box      = "horizontal",
            legend.title    = element_text(size = 10), 
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()
            ) +
      labs(fill = if(y == "CGI") {""} else {"Log2(normalized counts + 1)"}) +
      guides(fill = guide_colourbar(title.position = "top", title.hjust = 0.5, barwidth = unit(4, "cm"), barheight = unit(0.30, "cm"), frame.colour = "black", ticks.colour = "black"))
  }) %>% cowplot::plot_grid(plotlist = ., nrow = 1, rel_widths = c(0.75, 0.3, 0.3, 1))
       
# HNES1 cells heatmap
HNES1_ht <- 
counts_in_promoters_full %>%
filter(group %in% c("HNES1 d0", "HNES1 d10")) %>%
select(c("gene_id", "RNA-Seq", "H3K4me3", "H3K27me3", "CGI", "group")) %>%
unique %>% 
left_join(promoters_category_by_group[, c("gene_id", "HNES1 d0", "HNES1 d10")] %>%
          arrange(`HNES1 d0`, `HNES1 d10`) %>% 
          tidyr::unite("cluster", `HNES1 d0`:`HNES1 d10`, sep = "_") %>% 
          mutate(cluster = factor(cluster, levels = unique(cluster))),
          by = "gene_id") %>%
  select(-gene_id) %>%
  mutate(group = factor(group, levels = c("HNES1 d0", "cR-H9-EOS d0", "HNES1 d1", "cR-H9-EOS d1", "HNES1 d2", "cR-H9-EOS d2", "HNES1 d3", "cR-H9-EOS d3", "HNES1 d7", "cR-H9-EOS d7", "HNES1 d10", "cR-H9-EOS d10", "HNES1 d20X", "cR-H9-EOS d20X", "HNES1 d20E", "cR-H9-EOS d20E", "H9-EOS")
  )) %>% 
  filter(group %in% c("HNES1 d0", "HNES1 d1", "HNES1 d2", "HNES1 d3", "HNES1 d7", "HNES1 d10")) %>% 
  group_by(group, cluster) %>% 
  summarise_all(.funs = ~ mean(.x, na.rm = T)) %>% 
  ungroup() %>%
  gather(seqtype, value, -group, -cluster) %>% 
  named_group_split(seqtype) %>%
  .[c("H3K4me3", "H3K27me3", "CGI", "RNA-Seq")] %>%
  map2(names(.),
       
      function(x, y){
      ggplot(x %>% na.omit, aes(group, fct_rev(cluster), fill = value)) +
      geom_tile(color = "black", size = 0.2) +
      scale_fill_gradientn(colours = if(y == "CGI") {rev(brewer.pal(9, "YlGnBu"))} else {rev(brewer.pal(11, "RdBu"))}) +
      theme_publication() +
      theme(axis.title.x    = element_blank(),
            axis.title.y    = element_blank(),
            axis.text.y     = if(y == "H3K4me3") {element_text(hjust = 1)} else {element_blank()},
            axis.ticks.y    = element_blank(),
            axis.text.x     = element_text(angle = 45, hjust = 1, vjust = 1),
            panel.border    = element_blank(),
            legend.position = "bottom",
            legend.box      = "horizontal",
            legend.title    = element_text(size = 10), 
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()
            ) +
      labs(fill = if(y == "CGI") {""} else {"Log2(normalized counts + 1)"}) +
      guides(fill = guide_colourbar(title.position = "top", title.hjust = 0.5, barwidth = unit(4, "cm"), barheight = unit(0.30, "cm"), frame.colour = "black", ticks.colour = "black"))
  }) %>% cowplot::plot_grid(plotlist = ., nrow = 1, rel_widths = c(0.75, 0.3, 0.3, 1))

```

# Correlation between bivalent promoters and expression
```{r}

svg(filename = "output/figures/03_condition_comparison/bivalency_4.svg", width = 9, height = 7)
cRH9EOS_ht
dev.off()

svg(filename = "output/figures/03_condition_comparison/bivalency_5.svg", width = 9, height = 7)
HNES1_ht
dev.off()

```

# Session information
```{r session_info, echo = FALSE}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.