---
title:    "Epigenetic dynamics during capacitation of na誰ve human pluripotent stem cells"
subtitle: "MOFA2 - Epigenetic changes by combined marks (Factor 1)"
author:   "Jo達o Pedro Agostinho de Sousa"
date:     "April 19, 2021"
output:
  html_document:
    df_print: paged
    toc: TRUE
    #toc_float: TRUE
    theme: default
editor_options:
  chunk_output_type: console
---

# Dependencies
```{r Packages, results = "hide", warning = FALSE}

purrr::walk(list.files("analysis/functions", pattern = "load_project_packages", full.names = T), source)

```

# Loading results from MOFA2 factor 1
```{r}

plot_data <- readRDS(file = "output/database/MOFA2_w_methylation_plot_data.rds")

# USER SET
userSets_signif <- 
  plot_data %>%
  filter(abs(value) > 0.5) %>% 
  select(feature, feature_type) %>%
  unique %>%
  named_group_split(feature_type) %>%
  map(~
        tidyr::separate(.x, feature, c("chr", "start", "end"), sep = ":") %>%
        makeGRangesFromDataFrame(keep.extra.columns = T)
      )

# UNIVERSE
model <- load_model("tmp/MOFA2_peaks_w_methylation.hdf5")

weights <-
  get_weights(model, as.data.frame = T) %>%
  mutate(factor = factor(factor, paste0("Factor", 1:3))) %>%
  mutate(view = factor(view, levels = rev(c("ATAC-seq","H3K4me3","H3K27ac","H3K4me1","H3K27me3","H3K9me3","Methylation"))))

universeSets <-
  weights %>%
  mutate(feature = str_remove(feature, "_.*")) %>% 
  filter(factor == "Factor1") %>% 
  select(feature, view) %>%
  unique %>%
  named_group_split(view) %>%
  map(~
        tidyr::separate(.x, feature, c("chr", "start", "end"), sep = ":") %>%
        makeGRangesFromDataFrame(keep.extra.columns = T)
      )

```

```{r}

#genome_annotation$`CpG islands`@seqinfo@genome[] <- "hg38"
seqlevelsStyle(genome_annotation$`CpG islands`) <- "UCSC"
#genome_annotation$`CpG islands` <- renameSeqlevels(genome_annotation$`CpG islands`, mapSeqlevels(seqlevels(genome_annotation$`CpG islands`), "UCSC"))

CpGdensity <- 
  Repitools::cpgDensityCalc(x = genome_annotation$`CpG islands`, organism = Hsapiens, verbose = TRUE) %>% 
  tibble(gene_id = names(.), CpG_density = .)

genome_annotation$`CpG islands`@seqinfo@genome[] <- "GRCh38"
#genome_annotation$`CpG islands` <- renameSeqlevels(genome_annotation$`CpG islands`, mapSeqlevels(seqlevels(genome_annotation$`CpG islands`), "Ensembl"))
seqlevelsStyle(genome_annotation$`CpG islands`) <- "Ensembl"



CGIs <- 
genome_annotation$`CpG islands` %>% as.data.frame() %>% mutate(CpGdensity = CpGdensity$CpG_density) %>% mutate(CpGdensity_corrected = CpGdensity/width) %>% mutate(overlap = FALSE) %>% makeGRangesFromDataFrame(keep.extra.columns = T) %>% .[seqnames(.) %in% 1:22,]

queryhits <- findOverlaps(CGIs, userSets_signif$`H3K27me3.Primed (10988)`) %>% as.data.frame() %>% pull(queryHits)
CGIs <- as.data.frame(CGIs)
CGIs[queryhits, c("overlap")] <- T

CGIs %>%
  ggplot(aes(CpGdensity_corrected, fill = overlap)) +
  geom_density(alpha = 0.3) +
  theme_publication() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

ks.test((CGIs %>% filter(overlap == T) %>% pull(CpGdensity_corrected)), (CGIs %>% filter(overlap == F) %>% pull(CpGdensity_corrected)), alternative = "less")




CGIs %>% ggplot(aes(width, CpGdensity_corrected)) + geom_point(size = 1, alpha = 0.5) + geom_smooth(method = "lm", formula = "y ~ x") + scale_x_log10()



CGIs %>% filter(overlap == T) %>% filter(CpGdensity_corrected > 0.12)

a <- subsetByOverlaps(genome_annotation$`Gene bodies`, (CGIs %>% filter(CpGdensity_corrected > 0.12) %>% makeGRangesFromDataFrame())) %>% .$gene_id %>% unique
b <- subsetByOverlaps(genome_annotation$`Gene bodies`, (CGIs %>% makeGRangesFromDataFrame())) %>% .$gene_id %>% unique


#a <- subsetByOverlaps(genome_annotation$`Gene bodies`, (CGIs %>% filter(width > 2000) %>% makeGRangesFromDataFrame())) %>% .$gene_id %>% unique
#b <- subsetByOverlaps(genome_annotation$`Gene bodies`, (CGIs %>% makeGRangesFromDataFrame())) %>% .$gene_id %>% unique

GOterms <- 
enrichGO(gene          = a,
         universe      = b,
         keyType       = "ENSEMBL",
         OrgDb         = org.Hs.eg.db,
         ont           = "BP",
         pAdjustMethod = "BH",
         qvalueCutoff  = 0.05,
         readable      = TRUE
         ) %>%
  clusterProfiler::simplify(cutoff = 0.7, by = "p.adjust", select_fun = min)

barplot(GOterms)

```


```{r}

# Subset of the annotation files
genome_annotation_subset <- 
  genome_annotation %>% .[c("Promoters", "Distal Na誰ve enhancers", "Distal Na誰ve super-enhancers", "Distal Primed enhancers", "Distal Primed super-enhancers", "Exons", "Introns", "Intergenic regions")] %>% GRangesList()


annotate_regions_by_frequency(genomic_regions             = (CGIs %>% filter(width > 2000) %>% makeGRangesFromDataFrame(keep.extra.columns = T)),
                              annotation_list             = genome_annotation_subset,
                              name                        = "CpG islands",
                              with_duplicated_annotations = T,
                              apply_GRanges_reduce        = F)

CGIs_gr <- CGIs %>% mutate(CGI_width = width) %>% makeGRangesFromDataFrame(keep.extra.columns = T)
seqlevelsStyle(CGIs_gr) <- "Ensembl"

ids <- subsetByOverlaps(genome_annotation$`Gene bodies`, CGIs_gr %>% .[.$CGI_width > 2000]) %>% as.data.frame() %>% pull(gene_id) %>% unique
idsu <- subsetByOverlaps(genome_annotation$`Gene bodies`, CGIs_gr) %>% as.data.frame() %>% pull(gene_id) %>% unique

CGIs_gr %>% as.data.frame() %>% arrange(dplyr::desc(CGI_width)) %>% mutate(rank = 1:nrow(.)) %>% ggplot(aes(rank, CGI_width)) + geom_hline(yintercept = 2000) + geom_vline(xintercept = 1276) + geom_line() + scale_y_log10()

require(biomaRt)

mart <- useMart('ENSEMBL_MART_ENSEMBL', host = 'useast.ensembl.org')
mart <- useDataset('hsapiens_gene_ensembl', mart)

annotLookup <- getBM(
  mart = mart,
  attributes = c(
    'hgnc_symbol',
    'ensembl_gene_id', 
    'description',
    'percentage_gene_gc_content'))

listAttributes(mart) %>% View

annotLookup %>% filter(ensembl_gene_id %in% ids) %>% pull(superfamily) %>% unique
annotLookup %>% filter(ensembl_gene_id %in% ids) %>% mutate(description = str_extract(description, "[^\\[]*")) %>% unique %>% View


install.packages("gprofiler2")
library(gprofiler2)

gostres <- gprofiler2::gost(query = ids, custom_bg = idsu, organism = "hsapiens", user_threshold = 0.05, significant = T, domain_scope = "custom_annotated", correction_method = "fdr")

gostres$result %>%
  filter(term_size < 1000) %>%
  as.data.frame() %>% 
  mutate(p_value = -log10(p_value)) %>%
  filter(p_value > 2) %>% 
  group_by(source) %>% 
  slice_max(order_by = p_value, n = 5) %>% 
  arrange(dplyr::desc(p_value)) %>% 
  mutate(term_name = factor(term_name, levels = unique(term_name))) %>% 
  ungroup %>% 
  ggplot(aes(term_name, p_value, fill = p_value)) + 
  geom_col(color = "black") + 
  facet_grid(~ source,scales = "free_x", space = "free_x") + 
  theme_publication() + 
  theme(axis.text.x      = element_text(hjust = 1, vjust = 1, angle = 45), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
  scale_fill_distiller(palette = "Reds", direction = -1) +
  labs(x = "", y = "-Log10(adjusted p-value)")

gprofiler2::gostplot(gostres)

```


# Session information
```{r session_info, echo = FALSE}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.